source/base/hdf5.cc

```diff
@@ -0,0 +1,1752 @@
+        }
+      else if (std::is_same<number, double>::value)
+        {
+          return std::make_shared<hid_t>(H5T_NATIVE_DOUBLE);
+        }
+      else if (std::is_same<number, int>::value)
+        {
+          return std::make_shared<hid_t>(H5T_NATIVE_INT);
+        }
+      else if (std::is_same<number, unsigned int>::value)
+        {
+          return std::make_shared<hid_t>(H5T_NATIVE_UINT);
+        }
+      else if (std::is_same<number, std::complex<float>>::value)
+        {
+          t_type  = std::shared_ptr<hid_t>(new hid_t, [](auto pointer) {
+            // Relase the HDF5 resource
+            H5Tclose(*pointer);
+            delete pointer;
+          });
+          *t_type = H5Tcreate(H5T_COMPOUND, sizeof(std::complex<float>));
+          //  The C++ standards committee agreed to mandate that the storage
+          //  format used for the std::complex type be binary-compatible with
+          //  the C99 type, i.e. an array T[2] with consecutive real [0] and
+          //  imaginary [1] parts.
+          H5Tinsert(*t_type, "r", 0, H5T_NATIVE_FLOAT);
+          H5Tinsert(*t_type, "i", sizeof(float), H5T_NATIVE_FLOAT);
+          return t_type;
+        }
+      else if (std::is_same<number, std::complex<double>>::value)
+        {
+          t_type  = std::shared_ptr<hid_t>(new hid_t, [](auto pointer) {
+            // Relase the HDF5 resource
+            H5Tclose(*pointer);
+            delete pointer;
+          });
+          *t_type = H5Tcreate(H5T_COMPOUND, sizeof(std::complex<double>));
+          //  The C++ standards committee agreed to mandate that the storage
+          //  format used for the std::complex type be binary-compatible with
+          //  the C99 type, i.e. an array T[2] with consecutive real [0] and
+          //  imaginary [1] parts.
+          H5Tinsert(*t_type, "r", 0, H5T_NATIVE_DOUBLE);
+          H5Tinsert(*t_type, "i", sizeof(double), H5T_NATIVE_DOUBLE);
+          return t_type;
+        }
+
+      // The function should not reach this point
+      Assert(false, ExcInternalError());
+      return t_type;
+    }
+
+
+
+    // Several HDF5 functions such as H5Screate_simple() require a
+    // one-dimensional array that specifies the size of each dimension of the
+    // container, see:
+    // https://support.hdfgroup.org/HDF5/doc1.8/RM/RM_H5S.html#Dataspace-CreateSimple
+    // The function get_container_dimensions returns a vector with the
+    // dimensions of the container.
+    // For a std::vector the function returns std::vector<hsize_t>{vector_size}
+    // For a Vector the function returns std::vector<hsize_t>{vector_size}
+    // For a FullMatrix the function returns std::vector<hsize_t>{rows, columns}
+    //
+    // Instead of using functions with std::enable_if, "constexpr if" from C++17
+    // could be used
+    template <typename number>
+    std::vector<hsize_t>
+    get_container_dimensions(const std::vector<number> &data)
```

Comment:
just drop the last sentence of the documentation -- I think it's not relevant any more.